rules/rules-integration-tests/src/shared/kotlinx/serialization/r8/DeletionChecks.kt

```diff
@@ -0,0 +1,213 @@
+     * Find R8 information about a method by its name.
+     * If there are several methods with name [name], [IllegalArgumentException] will be thrown.
+     * If there is no method with name [name], [IllegalArgumentException] will be thrown.
+     */
+    fun findMethod(name: String): R8Result {
+        val methods = clazz.methods.values.filter { it.name == name }
+        if (methods.isEmpty()) {
+            throw IllegalArgumentException("Method with name '$name' not found in class '${clazz.originalName}'")
+        } else if (methods.size > 1) {
+            throw IllegalArgumentException("Several methods with name '$name' found in class '${clazz.originalName}'")
+        } else {
+            return methods.single().let { R8Result(it.name != it.obfuscatedName, it.removed) }
+        }
+    }
+}
+
+
+data class ClassEntry(
+    val originalName: String,
+    val obfuscatedName: String,
+    var removed: Boolean = false,
+    val fields: MutableMap<String, FieldEntry> = mutableMapOf(),
+    val methods: MutableMap<String, MethodEntry> = mutableMapOf()
+)
+
+data class FieldEntry(
+    val name: String,
+    val type: String,
+    val obfuscatedName: String,
+    var removed: Boolean = false
+)
+
+data class MethodEntry(
+    val name: String,
+    val returnType: String,
+    val descriptor: String,
+    val obfuscatedName: String,
+    var removed: Boolean = false
+)
+
+
+open class R8Result(
+    val isObfuscated: Boolean,
+    val isShrunk: Boolean
+)
+
+private fun parseR8Output(mappingFile: File, usageFile: File): Map<String, ClassEntry> {
+    val classMap = mutableMapOf<String, ClassEntry>()
+    var currentClass: ClassEntry? = null
+
+    // process mapping.txt
+    val classRegex = Regex("""^(\S+) -> (\S+):$""")
+    val methodRegex = Regex("""^(?:(\d+):\d+:)?(\S+)\s+([^\s\\(]+)(\(.*\))?:\d+(:\d+)? -> (\S+)$""")
+    val fieldRegex = Regex("""^(\S+)\s+(\S+)\s+->\s+(\S+)$""")
+
+    mappingFile.forEachLine { raw ->
+        val line = raw.trim()
+        if (line.startsWith("#") || line.isEmpty()) return@forEachLine
+
+        classRegex.matchEntire(line)?.let {
+            val (original, obfuscated) = it.destructured
+            currentClass = ClassEntry(originalName = original, obfuscatedName = obfuscated)
+            classMap[original] = currentClass
+            return@forEachLine
+        }
+
+        methodRegex.matchEntire(line)?.let {
+            val current = currentClass ?: throw IllegalStateException("No current class")
+
+            val (num, returnType, name, desc, _, obfuscated) = it.destructured
+            if (num.isNotEmpty()) {
+                val signature = name + desc
+                val existed = current.methods[signature]
+                if (existed == null) {
+                    current.methods[signature] = MethodEntry(name, returnType, desc, obfuscated)
+                } else {
+                    if (existed.obfuscatedName != obfuscated) {
+                        if (obfuscated == name) {
```

Comment:
Please add comment explaining the logic of "If method saved its original name in one of the obfuscations, we use original, otherwise we use either of obfuscated"